### PR TITLE
Fix various problems with Chip delete button.

### LINKED ISSUE #42455: Chip delete button tap target is too small.
Now that `InkResponse` has a focus node in it, the delete button has a separate semantics node, which causes it to appear in the `examples/flutter_gallery/accessibility_test.dart` search for nodes to measure, and it's coming up as too small: it's 18x18 instead of 48x48, since the tap target is only the delete icon itself.

Fixing this will probably require wrapping the delete icon in a larger gesture detector above the `InkResponse`.

### LINKED ISSUE #43291: Chip delete button tap target is always 1/3 the width of the whole chip.
# Description
- On a chip with a delete button, the chip considers *any* tap on the rightmost 1/3 of the chip as a tap on the delete button. (On an RTL chip, it is the leftmost 1/3.)
- This tap target could be too large (if the chip is long) or too small (if the chip is short).
- To implement it correctly, the tap target for the delete button should depend *solely* on the location and size of the delete button, **not** the size of the entire chip. For example, we can let the tap target be a 48×48 square centered on the delete button. 

# Possible related issues
This issue may be related to the following issue:
- Chip delete button tap target is too small. #42455

# Steps to reproduce
1. Create a new flutter app.
2. Copy the code below and save it as `main.dart` for the app.
3. Run the app.
4. You will see a chip with a long text label and a delete button.
5. You will also see a text message below the chip.
6. When you click on different parts of the chip, the text message on the screen will show you which part of the chip (label vs. delete button) is clicked.
7. If you click on the word "chip" in the label (which is very near the delete button), the message will say "Delete button pressed" instead of "Label pressed".

# Code

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _messageCount = 0;
  String _onScreenMessage = 'Press the chip in different positions.';

  String get currentReport => '[$_messageCount] $_onScreenMessage';

  void _report(String report) {
    setState(() {
      _messageCount ++;
      _onScreenMessage = report;
    });
    print(currentReport);
  }

  @override
  Widget build(BuildContext context) {
    return Material(
      child: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            RawChip(
              label: Text("A supercalifragilisticexpialidocious chip."),
              onPressed: () {_report('Label pressed.');},
              onDeleted: () {_report('Delete button pressed.');},
              deleteIcon: Icon(Icons.close),
            ),
            Text(
              currentReport,
              style: Theme.of(context).textTheme.subtitle,
            ),
          ],
        ),
      ),
    );
  }
}
```

### LINKED ISSUE #43534: [Chip] Chip with onDeleted but without onPressed has no ripple when tapping delete button.
# Description

On a chip with a delete button but without `onPressed`, there is currently no ripple on the delete button when it is pressed.

### Behavior before merging #42779
- Tapping on a chip with only `onPressed` creates one ripple;
- Tapping on a chip with `onPressed` and `onDeleted` on the delete icon creates two ripples, one for the whole chip and one for the delete icon; **(unexpected)**
- Tapping on a chip with only `onDeleted` creates one ripple;
- Tapping on a chip with neither `onPressed` nor `onDeleted` creates no ripples.

(Highlights are not counted as ripples.)

The pull request #42779 fixes the issue b/140930220; after #42779, there is only one ripple when the delete icon is tapped on a chip with `onPressed` and `onDeleted`. But this pull request creates the following behavior:

### Behavior after merging #42779 (current master)
- Tapping on a chip with only `onPressed` creates one ripple;
- Tapping on a chip with `onPressed` and `onDeleted` on the delete icon creates one ripple for the delete icon only;
- Tapping on a chip with only `onDeleted` creates no ripples; **(unexpected)**
- Tapping on a chip with neither `onPressed` nor `onDeleted` creates no ripples.

(Tapping the delete button still calls `onDeleted` as expected. Holding down the delete button still shows a tooltip.)

### Expected behavior
- Tapping on a chip with only `onPressed` creates one ripple;
- Tapping on a chip with `onPressed` and `onDeleted` on the delete icon creates one ripple for the delete icon only;
- Tapping on a chip with only `onDeleted` creates one ripple for the delete icon only;
- Tapping on a chip with neither `onPressed` nor `onDeleted` creates no ripples.

Clarifications are needed for the following behavior:
- When a chip with a delete button is pressed (on the delete button), should the whole chip be **highlighted**?
- When a chip with a delete button is pressed (on the delete button), should the whole chip be elevated?

# Related issues
- b/140930220
- This issue is blocking PR #41463 (has caused #41463 to be reverted).

# Steps to reproduce

1. Create a new flutter project and save the following code as `main.dart`.
2. Run the project (with flutter on branch `master`).
3. Press and hold different parts of each chip to see how the ripples behave.

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            RawChip(
                label: Text('Chip with onPressed'),
                onPressed: () {},
            ),
            RawChip(
                label: Text('Chip with onPressed and onDeleted'),
                onDeleted: () {},
                onPressed: () {},
            ),
            RawChip(
                label: Text('Chip with onDeleted only'),
                onDeleted: () {},
            ),
            RawChip(
                label: Text('Inactive chip'),
            ),
          ],
        ),
      ),
    );
  }
}
```

# Screenshots
- Test app
home_screen
- Tapping first chip
tap_1
- Tapping second chip
tap_2
- Tapping delete icon on second chip (master)
tap_2_delete_button
- Tapping delete icon on second chip (previous version)
previous_doubleripple
- Tapping delete icon on third chip (previous version)
previous_ripple

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This fixes several problems with the delete button on `Chip` widgets:

- Before this, the hit area for the delete button was always 1/3 of the chip size, leading to hit areas that are too wide on large chips, and too small on small chips. This modifies that so that on large chips, it takes up to 28 pixels to the left (in LTR locales, right on RTL) of the center of the delete button (basically a 48x48 square around the delete button, but clipped by the Chip), but is at a minimum up to the left edge of the delete button of the chip on small chips.
- The ripple on the delete button was not being fired in all situations, and when it was, it was clipped to the 16x16 delete icon, making it almost invisible. Now it expands towards the center of the delete button, and stops just short of the edge of the chip.
- The delete button didn't have a focus node, so couldn't be accessed via the keyboard focus traversal. Now it has a focus node again.
- Fixes a small iteration issue in the mock canvas that was throwing an exception instead of showing where the draw operations differed.

This change does not revert to having two ripples for a delete press: there is still only one, since the hit testing separates the delete button InkWell from the main chip (i.e. it doesn't regress b/140930220).

Before (clicking on each active area of the chip):

https://user-images.githubusercontent.com/8867023/134395547-aa9feee8-4014-4065-a09f-ee7b77e98a77.mp4

After (clicking on each active area of the chip):

https://user-images.githubusercontent.com/8867023/134395349-35080575-c2ac-4630-8ada-73612bf27760.mp4

## Related Issues
 - Fixes https://github.com/flutter/flutter/issues/42455 It is still too small, but getting closer: it at least tries to be 24 pixels on the left side, plus the distance from the center of the chip to the right side of the chip, so closer, but not there yet (about 36 pixels wide by 32 pixels high on most chips). I think this is probably the best we can do without expanding outside of the size of the chip (which would increase the size of the Material, making it visually different).
 - Fixes https://github.com/flutter/flutter/issues/43291 Delete button is no longer a fixed ratio.
 - Fixes https://github.com/flutter/flutter/issues/43534 Chip with no `onPressed`, but with `onDelete` now has a ripple.

## Tests
 - Added a test for the delete button taking up the right amount of the chip.
 - Updated tests to allow for more than one InkWell, but still check for a single ripple.
 - Added a test to make sure the delete button is focusable.
 - Added a test to make sure that chips without `onPressed` set, but with `onDelete` set still ripple.

### CODE REVIEW

File: packages/flutter/lib/src/material/chip.dart [hunk 1]
```diff
@@ -1806,15 +1802,12 @@ class _RawChipState extends State<RawChip> with MaterialStateMixin, TickerProvid
...
-          key: deleteIconKey,
-          behavior: HitTestBehavior.opaque,
-          onTap: widget.isEnabled
-            ? () {
-                Feedback.forTap(context);
-                widget.onDeleted!();
-              }
-            : null,
+        InkWell(
+          // Radius should be slightly less than the full size of the chip.
```
Comment (darrenaustin) [inline packages/flutter/lib/src/material/chip.dart:1806]: Why "slightly less"? Do we just want to make sure it doesn't paint outside of the chip?

File: packages/flutter/lib/src/material/chip.dart [hunk 2]
```diff
@@ -2949,61 +2932,56 @@ class _LocationAwareInkRippleFactory extends InteractiveInkFeatureFactory {
...
+  // width of the tappable region within the chip, not outside of the chip, which is
+  // handled elsewhere. Also because delete buttons aren't specified to be used on
+  // touch devices, only desktop devices.
+  final double accessibleDeleteButtonWidth =
+  math.max(
+    deleteButtonSize.width,
+    math.min(
+      deflatedSize.width * 0.5,
+      24.0 + deleteButtonSize.width / 2.0,
+    ),
```
Comment (darrenaustin) [inline packages/flutter/lib/src/material/chip.dart:2979]: Really small nit: this should be indented.

File: packages/flutter/test/material/chip_test.dart [hunk 3]
```diff
@@ -3212,7 +3346,7 @@ void main() {
   });
 
   testWidgets('intrinsicHeight implementation meets constraints', (WidgetTester tester) async {
-    // Regression text for https://github.com/flutter/flutter/issues/49478.
+    // Regression test for https://github.com/flutter/flutter/issues/49478.
```
Comment (darrenaustin) [inline packages/flutter/test/material/chip_test.dart:3349]: Nice 😄.

**Review by darrenaustin — APPROVED** [review]
LGTM. Just a couple of small comments. Nice cleanup.

File: packages/flutter/lib/src/material/chip.dart (hunk 1 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/chip.dart:1806]: Yeah, I could make it half, but then the last part of the splash paints outside the chip.  Aesthetically, I didn't think it looked right.  I'm happy to make it 0.5, though if you think it's better to be agnostic about it.

File: packages/flutter/lib/src/material/chip.dart (hunk 2 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/chip.dart:2979]: Sorry, I don't see what you mean. It's in line with the `math.min` line, which is what it matches.

File: packages/flutter/lib/src/material/chip.dart (hunk 1 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/material/chip.dart:1806]: No that's fine. I was just curious about the reasoning behind it. 0.45 looked good to me.

File: packages/flutter/lib/src/material/chip.dart (hunk 2 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/material/chip.dart:2979]: I meant that `math.max` and it's parameters should be indented on level under the `accssibleDeleteButtonWidth =` line. Again this is a total nit.

Comment (gspencergoog) [reply packages/flutter/lib/src/material/chip.dart:2979]: Oh, duh. OK, fixed.

**Pull request merged**